File: src/main/java/com/severett/chargerapp/model/ChargingSession.java
package com.severett.chargerapp.model;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

public final class ChargingSession {

    private final UUID id;
    private final String stationId;
    private Instant startedAt;
    private Instant stoppedAt;
    private Status status;

    public ChargingSession(UUID id, String stationId, Instant startedAt) {
        this.id = id;
        this.stationId = stationId;
        this.startedAt = startedAt;
        this.status = Status.IN_PROGRESS;
    }

    public UUID getId() {
        return id;
    }

    public String getStationId() {
        return stationId;
    }

    public Status getStatus() {
        return status;
    }

    public Instant getStartedAt() {
        return startedAt;
    }

    public Instant getStoppedAt() {
        return stoppedAt;
    }

    public void stopCharging(Instant stoppedAt) {
        status = Status.FINISHED;
        this.stoppedAt = stoppedAt;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ChargingSession that = (ChargingSession) o;
        return id.equals(that.id) &&
                stationId.equals(that.stationId) &&
                startedAt.equals(that.startedAt) &&
                Objects.equals(stoppedAt, that.stoppedAt) &&
                status == that.status;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, stationId, startedAt, stoppedAt, status);
    }
}

File: src/main/kotlin/com/severett/chargerapp/model/ChargingSession.kt
package com.severett.chargerapp.model

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonInclude
import com.severett.chargerapp.util.NoArgs
import java.time.Instant
import java.util.*

@NoArgs
class ChargingSession(
    var id: UUID,
    var stationId: String,
    @field:JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS", timezone = "UTC")
    var startedAt: Instant
) {

    var status = Status.IN_PROGRESS

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS", timezone = "UTC")
    var stoppedAt: Instant? = null

    fun stopCharging(stoppedAt: Instant) {
        status = Status.FINISHED
        this.stoppedAt = stoppedAt
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ChargingSession

        if (id != other.id) return false
        if (stationId != other.stationId) return false
        if (startedAt != other.startedAt) return false
        if (status != other.status) return false
        if (stoppedAt != other.stoppedAt) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + stationId.hashCode()
        result = 31 * result + startedAt.hashCode()
        result = 31 * result + status.hashCode()
        result = 31 * result + (stoppedAt?.hashCode() ?: 0)
        return result
    }

}
